### PR TITLE
Sample App for Dabl

### PR DESCRIPTION
**Pull request by gaborh-da**
- create a DIT package
- allow parties to be specified AFTER building the UI (via parties.json uploader / button)

### CODE REVIEW

**Review by tamaskalcza-da — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -51,3 +51,11 @@ dabl: dabl-ui
 
 dabl-ui: build
 	$(MAKE) -C ui dabl-ui LEDGER_ID=$(LEDGER_ID)
+
+DAR_FILES=$(shell find . -type f -name "*.dar")
+
+package: dabl
```
Comment (richardkapolnai-da) [inline Makefile:57]: not too important, merely a suggestion (feel free to ignore this)
it's just `package` sounds like a useful goal to our users. however, it's internal, we should frighten them a bit not to use this.
```suggestion
dabl-package: dabl
```

Comment (gaborh-da) [reply Makefile]: Done

File: Makefile [hunk 2]
```diff
@@ -50,4 +50,14 @@ reset-damltest: reset-damldir
...
 
 dabl-ui: build
-	$(MAKE) -C ui dabl-ui LEDGER_ID=$(LEDGER_ID)
+	$(MAKE) -C ui dabl-package LEDGER_ID=$(LEDGER_ID)
+
+DAR_FILES=$(shell find . -type f -name "*.dar")
+
+dit-package: build
+	$(MAKE) -C ui cbdc-ui.zip LEDGER_ID="dummy-dit"
+	ls .
```
Comment (richardkapolnai-da) [inline Makefile:59]: perhaps remove this line?
```suggestion
```

**Review by richardkapolnai-da — APPROVED** [review]

File: ui/src/config.ts [hunk 3]
```diff
@@ -5,8 +5,19 @@
 
 const credentials = require(process.env.REACT_APP_CONFIG_FILE as string);
 
-export const credentialsMap = credentials;
+export const preloadedCredentialsMap = credentials;
 
 export const isLocal = process.env.REACT_APP_IS_LOCAL === "true";
 
-export const httpBaseUrl = process.env.REACT_APP_HTTP_BASE_URL as string;
```
Comment (gaborh-da) [inline ui/src/config.ts:12]: You cannot **pre**configure the ledgerId / httpBaseUrl if you try to create a dit file. You just don't know it. That's why I included this from the DA Marketplace app.

I am open to suggestions.

Comment (gaborh-da) [reply ui/src/config.ts:12]: Let's use `getLedgerId` and throw exception when REACT_APP_LEDGER_ID is null.
REACT_APP_LEDGER_ID should be defined (it is not yet).

File: ui/src/Credentials.ts [hunk 4]
```diff
@@ -41,3 +42,49 @@ export const computeCredentials = (displayName: string): Credentials => {
...
+  return { partyId: partyIdentifier, token, ledgerId }
+}
+
+const PARTIES_STORAGE_KEY = 'imported_parties';
+
+export function storeParties(parties: PartyDetails[]): void {
+    localStorage.setItem(PARTIES_STORAGE_KEY, JSON.stringify(parties));
+}
+
+export function retrieveParties(validateParties: boolean = true): PartyDetails[] | undefined {
```
Comment (richardkapolnai-da) [inline ui/src/Credentials.ts:70]: if possible, don't export these.
perhaps provide a  `export function ReloadButton() { return (<DablPartiesInput ...`

**Comment by gaborh-da** [comment]
Screenshot from 2021-04-16 17-43-57

File: ui/src/components/Molecules/BtnActions/BtnActions.tsx [hunk 5]
```diff
@@ -73,9 +76,28 @@ const BtnActions: React.FC<BtnActionsProps> = (props) => {
...
+
     return (
         <div className={`${styles.actions}`}>
-            {buttons.map(button=><TextButton textBtnIconStyle={`${isDropdownShown && button.text==="SECTIONS"?styles._activeDropdown:null}`} key={Math.random()} text={button.text} icon={button.icon?button.icon:null} handleAction={button.text==="RELOAD"?handleLedgerReset:handleToggleDropdown}/>)}
+            {buttons.map(
+                button =>
+                  "LOAD PARTY FILE" === button.text
+                  ? <DablPartiesInput
+                        ledgerId={ledgerId}
+                        onError={error => console.log(error)}
```
Comment (richardkapolnai-da) [inline ui/src/components/Molecules/BtnActions/BtnActions.tsx:92]: wondering if we want to tolerate these errors and not throw an exception instead?

File: ui/src/Credentials.ts [hunk 6]
```diff
@@ -41,3 +42,49 @@ export const computeCredentials = (displayName: string): Credentials => {
...
+    const partiesJson = localStorage.getItem(PARTIES_STORAGE_KEY);
+
+    if (!partiesJson) {
+        return undefined;
+    }
+
+    const [ parties, error ] = convertPartiesJson(partiesJson, ledgerId, validateParties);
+
+    if (error) {
+        console.warn("Tried to load an invalid parties file from cache.", error);
```
Comment (richardkapolnai-da) [inline ui/src/Credentials.ts:80]: wondering if we want to tolerate these errors and not throw an exception instead?

File: ui/src/Credentials.ts (hunk 4 above)
Comment (gaborh-da) [reply ui/src/Credentials.ts:70]: good idea

File: ui/src/components/Molecules/BtnActions/BtnActions.tsx (hunk 5 above)
Comment (gaborh-da) [reply ui/src/components/Molecules/BtnActions/BtnActions.tsx:92]: yes we can do, I just wanted to take care of the happy path first

File: ui/src/Credentials.ts (hunk 6 above)
Comment (gaborh-da) [reply ui/src/Credentials.ts]: done

File: ui/src/components/Molecules/BtnActions/BtnActions.tsx (hunk 5 above)
Comment (gaborh-da) [reply ui/src/components/Molecules/BtnActions/BtnActions.tsx]: done

**Comment by gaborh-da** [comment]
I merged back main to follow the changes.

**Comment by gaborh-da** [comment]
The goal here is:
- to build a dit file (even if dit is not yet supported, we can build it)
- to make the UI support deployment without knowing parties beforehand
- to make the UI support upload of a parties.json file (to initialize it, like in DA Marketplace)

**Comment by gaborh-da** [comment]
Most of the ideas here are:
- from @tamaskalcza-da 's previous PRs
- from DA Marketplace / Dabl Team

**Comment by gaborh-da** [comment]
We said we do not need this. Closing.

**Pull request closed without merging**